v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -329,73 +472,123 @@ async def connect(self):
 
             else:
                 logger.debug("Already connected!")
 
+    async def _do_connect(self):
+        """Connect, start network loop, and wait for response"""
+
+        # Paho Connect
+        logger.debug("Attempting connect using port {}...".format(self._port))
+        try:
+            rc = await self._loop.run_in_executor(
+                None,
+                functools.partial(
+                    self._mqtt_client.connect,
+                    host=self._hostname,
+                    port=self._port,
+                    keepalive=self._keep_alive,
+                ),
+            )
+            rc_msg = mqtt.error_string(rc)
+            logger.debug("Connect returned rc {} - {}".format(rc, rc_msg))
+        # TODO: more specialization of errors to indicate which are/aren't retryable
+        except asyncio.CancelledError:
+            # Handled in outer method
+            raise
+        except Exception as e:
+            raise MQTTConnectionFailedError(message="Failure in Paho .connect()") from e
+
+        if rc != mqtt.MQTT_ERR_SUCCESS:
+            # NOTE: This block should probably never execute. Paho's .connect() is
+            # supposed to only return success or raise an exception.
+            logger.warning("Unexpected rc {} from Paho .connect()".format(rc))
+            # MQTTConnectionFailedError expects a connack rc, but this is a regular rc.
+            # So chain a regular mqtt exception into a connection mqtt exception.
+            try:
+                raise MQTTError(rc=rc)
+            except MQTTError as e:
+                raise MQTTConnectionFailedError(message="Unexpected Paho .connect() rc") from e
+
+        # Start Paho network loop, and store the task. This task will complete upon disconnect
+        # whether due to invocation of .disconnect(), or an unexpected network drop.
+        #
+        # NOTE: If the connect attempt is cancelled, the network loop cannot be stopped.
+        # This is fine though, as it'll be stopped on the next disconnect.
+        # However, this does introduce a case where the network loop may already be running
+        # during a connect attempt, so make sure it isn't before trying to start it again.
+        #
+        # NOTE: This MUST be called after connecting - loop_forever requires a socket to have been
+        # already established. This is not true of other network loop APIs, but it is true of this
+        # one.
+        if not self._network_loop_running():
+            logger.debug("Starting Paho network loop")
+            self._network_loop = self._loop.run_in_executor(None, self._mqtt_client.loop_forever)
+        else:
+            logger.debug(
+                "Paho network loop was already running. Likely due to a previous cancellation."
+            )
+
+        # The result of the CONNACK is received via this future stored on the lock
+        logger.debug("Waiting for connect response...")
+        rc = await self._pending_connect
```

Comment:
>  rc = await self._pending_connect [](http://example.com/codeflow?start=6&length=34)

Returning an RC isn't really very pythonic. Wouldn't it make more sense to have `await _self._pending_connect` raise an error if something goes wrong? 

Also, the comment above on "future stored on this lock" is outdated.

Reply:
Good catch on the lock stuff.

As far as the rc, I disagree - we use rc throughout the mqtt_client implementation, I see no reason to inject a spurious layer of abstraction between returning the result from the callback. As you can see above on the .connect() invocation, we are already directly using rcs in this method, so it's more consistent.

My view is that within this class that directly deals with Paho, we should be using rcs, since they are returned by Paho, and we have to just deal with that. But they should never, ever make it outside of the MQTTClient.